File: src/com/intellij/idea/plugin/hybris/settings/HybrisRemoteConnectionSettings.java
package com.intellij.idea.plugin.hybris.settings;

import org.apache.commons.lang3.StringUtils;

import java.io.Serializable;
import java.util.Objects;

public class HybrisRemoteConnectionSettings implements Serializable {

    public enum Type {Hybris, SOLR}

    protected String uuid;
    protected String displayName;
    protected String hostIP;
    protected String port;
    protected String solrWebroot;
    protected String adminLogin;
    protected String adminPassword;
    protected String hacWebroot;
    protected String hacLogin;
    protected String hacPassword;
    protected String generatedURL;
    protected Type type;
    protected boolean isHacSsl;
    protected boolean isSolrSsl;

    public boolean isSolrSsl() {
        return isSolrSsl;
    }

    public void setSolrSsl(final boolean solrSsl) {
        isSolrSsl = solrSsl;
    }

    public boolean isHacSsl() {
        return isHacSsl;
    }

    public void setHacSsl(final boolean hacSsl) {
        isHacSsl = hacSsl;
    }

    public String getUuid() {
        return uuid;
    }

    public void setUuid(final String uuid) {
        this.uuid = uuid;
    }

    public String getDisplayName() {
        return displayName;
    }

    public void setDisplayName(final String displayName) {
        this.displayName = displayName;
    }

    public String getHostIP() {
        return hostIP;
    }

    public void setHostIP(final String hostIP) {
        this.hostIP = hostIP;
    }

    public String getPort() {
        return port;
    }

    public void setPort(final String port) {
        this.port = port;
    }

    public String getSolrWebroot() {
        return solrWebroot;
    }

    public void setSolrWebroot(final String solrWebroot) {
        this.solrWebroot = solrWebroot;
    }

    public String getAdminLogin() {
        return adminLogin;
    }

    public void setAdminLogin(final String adminLogin) {
        this.adminLogin = adminLogin;
    }

    public String getAdminPassword() {
        return adminPassword;
    }

    public void setAdminPassword(final String adminPassword) {
        this.adminPassword = adminPassword;
    }

    public String getHacWebroot() {
        return hacWebroot;
    }

    public void setHacWebroot(final String hacWebroot) {
        this.hacWebroot = hacWebroot;
    }

    public String getHacLogin() {
        return hacLogin;
    }

    public void setHacLogin(final String hacLogin) {
        this.hacLogin = hacLogin;
    }

    public String getHacPassword() {
        return hacPassword;
    }

    public void setHacPassword(final String hacPassword) {
        this.hacPassword = hacPassword;
    }

    public String getGeneratedURL() {
        return generatedURL;
    }

    public void setGeneratedURL(final String generatedURL) {
        this.generatedURL = generatedURL;
    }

    public Type getType() {
        return type;
    }

    public void setType(final Type type) {
        this.type = type;
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        final HybrisRemoteConnectionSettings that = (HybrisRemoteConnectionSettings) o;
        return Objects.equals(getUuid(), that.getUuid());
    }

    @Override
    public int hashCode() {
        return Objects.hash(getUuid());
    }

    @Override
    public String toString() {
        if (!StringUtils.isBlank(getDisplayName())) {
            return getDisplayName();
        }
        return getGeneratedURL();
    }
}

File: src/com/intellij/idea/plugin/hybris/settings/HybrisRemoteConnectionSettings.kt
/*
 * This file is part of "SAP Commerce Developers Toolset" plugin for Intellij IDEA.
 * Copyright (C) 2019-2023 EPAM Systems <hybrisideaplugin@epam.com> and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.intellij.idea.plugin.hybris.settings

import com.intellij.credentialStore.CredentialAttributes
import com.intellij.ide.passwordSafe.PasswordSafe
import com.intellij.idea.plugin.hybris.common.HybrisConstants
import com.intellij.idea.plugin.hybris.tools.remote.RemoteConnectionScope
import com.intellij.idea.plugin.hybris.tools.remote.RemoteConnectionType
import com.intellij.idea.plugin.hybris.tools.remote.RemoteConnectionUtil
import com.intellij.openapi.components.BaseState
import com.intellij.util.xmlb.Accessor

class HybrisRemoteConnectionSettings : BaseState(), Comparable<HybrisRemoteConnectionSettings> {

    var uuid by string(null)
    var displayName by string(null)
    var scope by property(RemoteConnectionScope.PROJECT_PERSONAL) { false }
    var type by property(RemoteConnectionType.Hybris) { false }
    var hostIP by string(HybrisConstants.DEFAULT_HOST_URL)
    var port by string(null)
    var isSsl by property(true)
    var sslProtocol by string(HybrisConstants.DEFAULT_SSL_PROTOCOL)
    var adminLogin by string(null)
    var adminPassword: String?
        set(value) {
            val credentialAttributes = CredentialAttributes("SAP CX - $uuid", hacLogin)
            PasswordSafe.instance.setPassword(credentialAttributes, value)
        }
        get() {
            val credentialAttributes = CredentialAttributes("SAP CX - $uuid", hacLogin)
            return PasswordSafe.instance.getPassword(credentialAttributes)
                ?: if (type == RemoteConnectionType.Hybris) "nimda"
                else "server123"
        }
    var solrWebroot by string("solr")
    var hacWebroot by string("")
    var hacLogin by string("admin")
    var hacPassword: String
        set(value) {
            val credentialAttributes = CredentialAttributes("SAP CX - $uuid", hacLogin)
            PasswordSafe.instance.setPassword(credentialAttributes, value)
        }
        get() {
            val credentialAttributes = CredentialAttributes("SAP CX - $uuid", hacLogin)
            return PasswordSafe.instance.getPassword(credentialAttributes)
                ?: "nimda"
        }
    val generatedURL: String
        get() {
            return when (type) {
                RemoteConnectionType.Hybris -> RemoteConnectionUtil.generateUrl(isSsl, hostIP, port, hacWebroot)
                RemoteConnectionType.SOLR -> RemoteConnectionUtil.generateUrl(isSsl, hostIP, port, solrWebroot)
            }
        }

    override fun toString() = (displayName
        ?.takeIf { it.isNotBlank() }
        ?: generatedURL
            .takeIf { it.isNotBlank() }
        ?: super.toString()
        )
        .let { scope.shortTitle + " : " + it }

    override fun accepts(accessor: Accessor, bean: Any) = accessor.name != "adminPassword"
        && accessor.name != "hacPassword"
        && accessor.name != "scope"

    override fun compareTo(other: HybrisRemoteConnectionSettings) = uuid
        ?.compareTo(other.uuid ?: "")
        ?: -1

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is HybrisRemoteConnectionSettings) return false

        if (uuid != other.uuid) return false

        return true
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + (uuid?.hashCode() ?: 0)
        return result
    }

}